src/syntax/ast.rs

```diff
@@ -5,101 +5,108 @@ pub struct Resource {
 pub enum Entry {
     Message(Message),
-    Comment(Comment),
     Section(Section),
+    Comment(Comment),
     Junk(JunkEntry),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Message {
+pub struct Pattern {
+    pub elements: Vec<Expression>,
+    pub quoted: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Expression {
+    Pattern(Pattern),
+    String(String),
+    Number(Number),
+    MessageReference(String),
+    ExternalArgument(String),
+    SelectExpression {
+        exp: Option<Box<Expression>>,
+        variants: Vec<Variant>,
+    },
+    AttributeExpression { id: Identifier, name: Identifier },
+    VariantExpression { id: Identifier, key: VariantKey },
+    CallExpression {
+        callee: Builtin,
+        args: Vec<Argument>,
+    },
+    Expression(Box<Expression>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Attribute {
     pub id: Identifier,
-    pub value: Option<Pattern>,
-    pub traits: Option<Vec<Member>>,
-    pub comment: Option<Comment>,
+    pub value: Pattern,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Pattern {
-    pub elements: Vec<PatternElement>,
-    pub quoted: bool,
+pub struct Variant {
+    pub key: VariantKey,
+    pub value: Pattern,
+    pub default: bool,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum PatternElement {
-    Text(String),
-    Placeable { expressions: Vec<Expression> },
+pub enum VariantKey {
+    Key(Keyword),
+    Number(Number),
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Identifier {
-    pub name: String,
+pub enum Argument {
+    Expression(Expression),
+    NamedArgument { name: Identifier, val: ArgValue },
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Key {
+pub enum ArgValue {
+    Number(Number),
+    String(String),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Identifier {
     pub name: String,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Number {
+pub struct Keyword {
     pub value: String,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Keyword {
-    pub ns: Option<Identifier>,
-    pub name: Key,
+pub struct Comment {
+    pub body: String,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum MemberKey {
-    Keyword(Keyword),
-    Number(Number),
+pub struct Section {
+    pub key: Keyword,
+    pub body: Vec<Entry>,
+    pub comment: Option<Comment>,
 }
 
 #[derive(Debug, PartialEq)]
-pub struct Member {
-    pub key: MemberKey,
-    pub value: Pattern,
-    pub default: bool,
+pub struct Number {
+    pub value: String,
 }
 
 #[derive(Debug, PartialEq)]
-pub enum Expression {
-    MessageReference { id: Identifier },
-    ExternalArgument { id: Identifier },
-    CallExpression {
-        callee: Identifier,
-        args: Vec<Expression>,
-    },
-    SelectExpression {
-        exp: Box<Expression>,
-        vars: Vec<Member>,
-    },
-    KeyValueArgument {
-        name: Identifier,
-        val: Box<Expression>,
-    },
-    Number(Number),
-    String(String),
-    Member {
-        obj: Box<Expression>,
-        key: MemberKey,
-    },
+pub struct Builtin {
```

Comment:
Maybe rename to `Function`? That's how we refer to them in the docs. Also, the ASDL defines the `name` field.